pallets/collator-selection/src/lib.rs

```diff
@@ -330,17 +330,46 @@ pub mod pallet {
 		pub fn account_id() -> T::AccountId {
 			T::PotId::get().into_account()
 		}
+		/// Removes a candidate if they exist and sends them back their deposit
+		fn try_remove_candidate(who: &T::AccountId) -> Result<usize, DispatchError> {
+			let current_count = <Candidates<T>>::try_mutate(|candidates| -> Result<usize, DispatchError> {
+				let index = candidates.iter().position(|candidate| candidate.who == *who).ok_or(Error::<T>::NotCandidate)?;
+				T::Currency::unreserve(&who, candidates[index].deposit);
+				candidates.remove(index);
+				Ok(candidates.len())
+			});
+			Self::deposit_event(Event::CandidateRemoved(who.clone()));
+			current_count
+		}
 
 		/// Assemble the current set of candidates and invulnerables into the next collator set.
 		///
 		/// This is done on the fly, as frequent as we are told to do so, as the session manager.
-		pub fn assemble_collators() -> Vec<T::AccountId> {
+		pub fn assemble_collators(candidates: Vec<T::AccountId>) -> Vec<T::AccountId> {
 			let mut collators = Self::invulnerables();
 			collators.extend(
-				Self::candidates().into_iter().map(|c| c.who).collect::<Vec<_>>(),
+				candidates.into_iter().collect::<Vec<_>>(),
 			);
 			collators
 		}
+		/// Kicks out and candidates that did not produce a block in the last session.
```

Comment:
```suggestion
		/// Kicks out and candidates that did not produce a block since the last call to this function. Given that this is called once per session, this essentially ensures that the each collator must author one block per session. 
```


and if we have more collators per session length? :D

Reply:
With the last commit this now can be a multiple of sessions and can be extended :)